File: Assets/Plugins/iOS/PlayPhoneMultiNet/unity/providers/MNAchievementsProviderUnity.m
//
//  MNAchievementsProviderUnity.m
//  PlayPhone SDK for Unity
//
//  Copyright (c) 2012 PlayPhone Inc. All rights reserved.
//

#import "MNDirect.h"
#import "MNAchievementsProvider.h"

#import "MNAchievementsProviderUnity.h"

@interface MNAchievementsProviderUnity : NSObject <MNAchievementsProviderDelegate>
+ (id)shared;
- (void)onGameAchievementListUpdated;
- (void)onPlayerAchievementUnlocked:(int)achievementId;
@end

static MNAchievementsProviderUnity *MNAchievementsProviderUnityInstance = nil;

@implementation MNAchievementsProviderUnity

+(id)shared {
    if (MNAchievementsProviderUnityInstance == nil) {
        MNAchievementsProviderUnityInstance = [[MNAchievementsProviderUnity alloc]init];
    }
    
    return MNAchievementsProviderUnityInstance;
}

- (void)onGameAchievementListUpdated {
    [MNUnity callUnityFunction:@"MNUM_onGameAchievementListUpdated" withParams:nil];
}
- (void)onPlayerAchievementUnlocked:(int)achievementId {
    [MNUnity callUnityFunction:@"MNUM_onPlayerAchievementUnlocked" withParams:[NSNumber numberWithInt:achievementId], nil];
}
@end

EXTERN_C
const char* _MNAchievementsProvider_GetGameAchievementsList() {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect achievementsProvider] getGameAchievementList]] UTF8String]);
}

EXTERN_C
const char* _MNAchievementsProvider_FindGameAchievementById(int achievementId) {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect achievementsProvider] findGameAchievementById:achievementId]] UTF8String]);
}

EXTERN_C
bool _MNAchievementsProvider_IsGameAchievementListNeedUpdate() {
    MARK;

    return [[MNDirect achievementsProvider] isGameAchievementListNeedUpdate];
}

EXTERN_C
void _MNAchievementsProvider_DoGameAchievementListUpdate() {
    MARK;

    [[MNDirect achievementsProvider] doGameAchievementListUpdate];
}

EXTERN_C
bool _MNAchievementsProvider_IsPlayerAchievementUnlocked(int achievementId) {
    MARK;

    return [[MNDirect achievementsProvider] isPlayerAchievementUnlocked:achievementId];
}

EXTERN_C
void _MNAchievementsProvider_UnlockPlayerAchievement(int achievementId) {
    MARK;

    [[MNDirect achievementsProvider] unlockPlayerAchievement:achievementId];
}

EXTERN_C
const char* _MNAchievementsProvider_GetPlayerAchievementsList() {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect achievementsProvider] getPlayerAchievementList]] UTF8String]);
}

EXTERN_C
const char* _MNAchievementsProvider_GetAchievementImageURL(int achievementId) {
    MARK;

    NSURL *resultUrl = [[MNDirect achievementsProvider] getAchievementImageURL:achievementId];

    if (resultUrl == nil) {
        return NULL;
    }

    return MakeStringCopy([[resultUrl absoluteString] UTF8String]);
}

EXTERN_C
bool _MNAchievementsProvider_RegisterEventHandler(){
    MARK;

    if ([MNDirect achievementsProvider] == nil) {
        return false;
    }

    [[MNDirect achievementsProvider] addDelegate:[MNAchievementsProviderUnity shared]];

    return true;
}

EXTERN_C
bool _MNAchievementsProvider_UnregisterEventHandler(){
    MARK;

    if ([MNDirect achievementsProvider] == nil) {
        return false;
    }

    [[MNDirect achievementsProvider] removeDelegate:[MNAchievementsProviderUnity shared]];

    return true;
}

File: Assets/Plugins/Android/src/com/playphone/multinet/unity/MNAchievementsProviderUnity.java
package com.playphone.multinet.unity;

import com.playphone.multinet.*;
import com.unity3d.player.UnityPlayer;

public class MNAchievementsProviderUnity {
    public static String getGameAchievementsList() {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getAchievementsProvider().getGameAchievementsList());
    }

    public static String findGameAchievementById(final int achievementId) {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getAchievementsProvider().findGameAchievementById(achievementId));
    }

    public static boolean isGameAchievementListNeedUpdate() {
        MNUnity.MARK();

        return MNDirect.getAchievementsProvider().isGameAchievementListNeedUpdate();
    }

    public static void doGameAchievementListUpdate() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getAchievementsProvider().doGameAchievementListUpdate();
            }
        });
    }

    public static boolean isPlayerAchievementUnlocked(final int achievementId) {
        MNUnity.MARK();

        return MNDirect.getAchievementsProvider().isPlayerAchievementUnlocked(achievementId);
    }

    public static void unlockPlayerAchievement(final int achievementId) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getAchievementsProvider().unlockPlayerAchievement(achievementId);
            }
        });
    }

    public static String getPlayerAchievementsList() {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getAchievementsProvider().getPlayerAchievementsList());
    }

    public static String getAchievementImageURL(final int achievementId) {
        MNUnity.MARK();

        return MNDirect.getAchievementsProvider().getAchievementImageURL(achievementId);
    }

    protected static class MNAchievementsProviderEventHandler implements com.playphone.multinet.providers.MNAchievementsProvider.IEventHandler {

        @Override
        public void onGameAchievementListUpdated() {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onGameAchievementListUpdated");
                }
            });
        }

        @Override
        public void onPlayerAchievementUnlocked(final int achievementId) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onPlayerAchievementUnlocked",achievementId);
                }
            });
        }
    }

    private static MNAchievementsProviderEventHandler registeredEventHandler = null;
    public static synchronized boolean registerEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getAchievementsProvider() == null) {
            MNUnity.ELog("MNAchievementsProvider is not ready. Use MNDirect\'s sessionReady event for adding your delegates.");
            return false;
        }

        if (registeredEventHandler == null) {
            registeredEventHandler = new MNAchievementsProviderUnity.MNAchievementsProviderEventHandler();
            MNDirect.getAchievementsProvider().addEventHandler(registeredEventHandler);
        }

        return true;
    }
    
    public static synchronized boolean unregisterEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getAchievementsProvider() == null) {
            return false;
        }

        if (registeredEventHandler != null) {
            MNDirect.getAchievementsProvider().removeEventHandler(registeredEventHandler);
            registeredEventHandler = null;
        }

        return true;
    }
}

